### PR TITLE
fix: route protocol-less binary pickles

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary
- Route security-relevant protocol-less binary pickle streams through pickle analysis even when suffixes are misleading.
- Preserve pathless and remote bounded-prefix candidates across 7z, Hugging Face, JFrog, and cloud routing.
- Fail closed with an explicit `pickle_routing_inconclusive` outcome when opcode, line, or prefix budgets prevent a decision.

## Validation
- Published commit: `0b493aa8cfc9a36f45f2d526b674896dd2edf723`
- Exact remote tree verified against the tested worktree.
- 79 focused pytest cases passed across core, file detection, 7z/nested dispatch, remote routing, and scanner selection.
- Scoped Ruff, format, mypy, and `git diff --check` passed.
- Full local suite intentionally not run; CI owns broad coverage.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
Critical review and focused QA complete.

Fixed several routing bypasses and one false-positive class:
- continue structurally past the 64 KiB content probe without loading attacker-sized operands, so large/split BINBYTES, BINSTRING, BINUNICODE, LONG4, BINUNICODE8, BINBYTES8, and BYTEARRAY8 values cannot hide a later gadget
- preserve detection across exact probe-boundary padding, open MARK state, truncated final STOP, and a later ordinary PROTO opcode
- recognize EXT1/EXT2/EXT4 and NEWOBJ as binary evidence
- require stronger evidence for incomplete extension-only prefixes so MessagePack EXT2 overlap remains clean
- preserve benign complete/truncated scalar, Flax, text, LightGBM, RKNN, RAR, protobuf, and PyTorch-binary routing negatives

Validation (focused only; no full suite):
- routing/filter modules: 389 passed, 12 skipped (optional ONNX unavailable)
- protocol-less core slice: 21 passed
- CVE-2025-10155 pickle tests: 18 passed
- scoped Ruff, format check, mypy, and `git diff --check`: clean

The prior benchmark report showed four noisy microsecond-level regressions with only +0.9% aggregate median; established binary-PROTO routes still return before the new structural fallback, and CI will rerun on this head.

Published fix: `1445ca9816d6e40e60d6255bcc4039cd35708419`.

File: modelaudit/utils/file/detection.py [hunk 1]
```diff
@@ -1577,6 +1624,201 @@ def _has_bounded_binary_pickle_security_signal(sample: bytes) -> bool:
...
+    opcode_count = 0
+    offset = 0
+    has_binary_opcode = False
+    has_security_opcode = False
+    has_pre_stop_security_opcode = False
+    try:
+        with file_path.open("rb") as handle:
+            while offset < file_size:
+                if opcode_count >= PROTO0_1_MAX_PROBE_OPCODES:
+                    return has_binary_opcode
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/file/detection.py]: Confirmed and reproduced. The validated local fix changes both structural budget exits (`opcode_count` and line-argument budget) from `has_binary_opcode` to `has_binary_opcode and has_pre_stop_security_opcode`, so repeated benign `SHORT_BINUNICODE`/`POP` padding cannot route as pickle. Regression: `test_scan_file_does_not_route_benign_binary_padding_past_opcode_budget`. Integrated local head `e48fcdc2`; 126 focused routing tests plus Ruff/format/mypy pass. Shell GitHub auth cannot publish, so @codex address that feedback with this exact gate and regression.

File: modelaudit/utils/file/detection.py [hunk 2]
```diff
@@ -3796,6 +4043,12 @@ def detect_format_from_magic_bytes(
 
     if file_path is not None:
         pickle_probe_sample = _read_pickle_probe_sample(file_path, file_size, magic16)
+        if _has_bounded_protocolless_binary_pickle_security_signal(
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/file/detection.py]: Confirmed and reproduced through `SevenZipScanner._probe_detected_format`. The validated local fix lets pathless `detect_format_from_magic_bytes` calls use `_looks_like_protocolless_binary_pickle_security_signal(magic16, sample_is_prefix=file_size > len(magic16))`, while filesystem-backed calls retain Flax overlap validation. Regression: `test_probe_detected_format_recognizes_protocolless_binary_pickle`. Integrated local head `e48fcdc2`; 126 focused routing tests plus Ruff/format/mypy pass. Shell GitHub auth cannot publish, so @codex address that feedback using this implementation.

**Comment by mldangelo-oai** [comment]
Critical review found the two unresolved routing issues were still present on exact head `1445ca98`, plus a related 7z false negative: the archive probe read a bounded 1 KiB prefix but only passed 16 bytes to the protocol-less binary-pickle detector, so padding could hide the security opcode.

I fixed these locally on top of current `main`:
- `30ae97cb` merges current `main` into the PR branch.
- `387236e2` requires a pre-STOP security opcode before fail-closed routing at parser budget exits, enables pathless protocol-less detection, and makes 7z inspect its full bounded prefix.
- Added malicious short/padded positives and benign opcode-padding negatives.

Focused validation:
- `22 passed` in the protocol-less/core routing selection
- `14 passed` in the 7z probe selection
- `148 passed` in related file-type routing tests
- targeted Ruff, format, mypy, and `git diff --check` clean

HTTPS push is blocked by the configured token (`Invalid username or token`). Please apply/publish `387236e2` after syncing current `main`; the unresolved threads should remain open until the fix is on the PR head.

@codex

File: modelaudit/utils/file/detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/file/detection.py]: Addressed in `0b493aa8`. Budget exhaustion no longer becomes a confirmed pickle route based on benign binary opcodes. It returns the explicit `pickle_routing_inconclusive` sentinel, producing an operationally visible incomplete scan without a security finding, cache entry, or exit-code 0. This also prevents an attacker from moving the gadget just past the budget to regain an `unknown` route.

File: modelaudit/utils/file/detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/file/detection.py]: Addressed in `0b493aa8`. The pathless API now inspects the caller's complete bounded prefix rather than only `magic16`, and carries whether that sample is truncated. 7z passes its 1 KiB probe; remote Hugging Face, JFrog, and cloud probes do the same. Delayed gadgets inside the prefix route to pickle, and candidates at the boundary fail closed instead of being skipped.

File: modelaudit/utils/file/detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/file/detection.py]: Addressed in `0b493aa8`. The structural probe is now tri-state: confirmed security signals route to pickle, complete benign streams stay unrouted, and opcode/line budget exhaustion returns `pickle_routing_inconclusive`. Core and nested dispatch fail closed with `success=False`, exit code 2, and no cache entry, avoiding both the original false positive and the late-gadget false negative. Added benign padding, late gadget, oversized line operand, cache, exit-code, and nested regressions.

File: modelaudit/utils/file/detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/utils/file/detection.py]: Addressed in `0b493aa8`. Pathless detection now accepts the full bounded pickle probe plus an explicit prefix-completeness flag. SevenZip, Hugging Face, JFrog, and cloud routing pass their full prefixes; confirmed delayed gadgets route to pickle, while a valid candidate reaching the prefix boundary is preserved as `pickle_routing_inconclusive` for fail-closed extraction/download and local analysis. Added delayed and boundary regressions for each routing path.
